client.go

```diff
@@ -1484,66 +1419,39 @@ func (c *Client) subscribeCmd(cmd *protocol.SubscribeRequest, rw *replyWriter) *
 			rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
 			return nil
 		}
-		parsedToken, err := jwt.ParseWithClaims(cmd.Token, &subscribeTokenClaims{}, jwtKeyFunc(config))
-		if parsedToken == nil && err != nil {
-			c.node.logger.log(newLogEntry(LogLevelInfo, "invalid subscription token", map[string]interface{}{"error": err.Error(), "client": c.uid, "user": c.UserID()}))
-			rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
-			return nil
-		}
-		if claims, ok := parsedToken.Claims.(*subscribeTokenClaims); ok && parsedToken.Valid {
-			tokenChannel = claims.Channel
-			tokenInfo = claims.Info
-			tokenB64info = claims.Base64Info
-			tokenClient = claims.Client
-			expireAt = claims.StandardClaims.ExpiresAt
-
-			if c.uid != tokenClient {
-				rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
-				return nil
-			}
-			if cmd.Channel != tokenChannel {
-				rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
+		var (
+			token     subscribeToken
+			errVerify error
+		)
+		if token, errVerify = c.node.verifySubscribeToken(cmd.Token); errVerify != nil {
+			if errVerify == errTokenExpired {
+				_ = rw.write(&protocol.Reply{Error: ErrorTokenExpired})
 				return nil
 			}
-
-			if len(tokenInfo) > 0 {
-				channelInfo = tokenInfo
+			if errVerify == errTokenInvalidInfo {
+				c.node.logger.log(newLogEntry(LogLevelInfo, "can not decode provided info", map[string]interface{}{"user": c.UserID(), "client": c.uid, "error": errVerify.Error()}))
+				return DisconnectBadRequest
 			}
-
-			if tokenB64info != "" {
-				byteInfo, err := base64.StdEncoding.DecodeString(tokenB64info)
-				if err != nil {
-					c.node.logger.log(newLogEntry(LogLevelInfo, "can not decode provided info from base64", map[string]interface{}{"user": c.UserID(), "client": c.uid, "error": err.Error()}))
-					return DisconnectBadRequest
-				}
-				channelInfo = byteInfo
-			}
-		} else {
-			if validationErr, ok := err.(*jwt.ValidationError); ok {
-				if validationErr.Errors == jwt.ValidationErrorExpired {
-					// The only problem with token is its expiration - no other errors set in bitfield.
-					rw.write(&protocol.Reply{Error: ErrorTokenExpired})
-					return nil
-				}
-				c.node.logger.log(newLogEntry(LogLevelInfo, "invalid subscription token", map[string]interface{}{"error": err.Error(), "client": c.uid, "user": c.UserID()}))
-				rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
-				return nil
-			}
-			c.node.logger.log(newLogEntry(LogLevelInfo, "invalid subscription token", map[string]interface{}{"error": err.Error(), "client": c.uid, "user": c.UserID()}))
+			c.node.logger.log(newLogEntry(LogLevelInfo, "invalid subscription token", map[string]interface{}{"error": errVerify.Error(), "client": c.uid, "user": c.UserID()}))
 			rw.write(&protocol.Reply{Error: ErrorPermissionDenied})
 			return nil
 		}
+		if c.uid != token.UserID {
```

Comment:
Major thing here: subscribeToken does not use `UserID` - it contains `Client` string - which is an unique id of connection generated on server side after connection established.

Reply:
@FZambia UserID is Client string maybe we can rename that to avoid mistakes 
![image](https://user-images.githubusercontent.com/26422482/73188467-e9610600-40f0-11ea-94ee-d0483de3ddce.png)

Reply:
Ah, yes - was not obvious, those are different things so `Client` is more correct anyway, thanks for fixing!